lnwallet/channel.go

```diff
@@ -4643,13 +4580,73 @@ func (lc *LightningChannel) computeView(view *HtlcView,
 	// channel constraints to the final commitment state. If any fee
 	// updates are found in the logs, the commitment fee rate should be
 	// changed, so we'll also set the feePerKw to this new value.
-	filteredHTLCView, err := lc.evaluateHTLCView(
-		view, &ourBalance, &theirBalance, nextHeight, whoseCommitChain,
-		updateState,
+	filteredHTLCView, uncommitted, deltas, err := lc.evaluateHTLCView(
+		view, whoseCommitChain, nextHeight,
 	)
 	if err != nil {
 		return 0, 0, 0, nil, err
 	}
+
+	// Add the balance deltas to the balances we got from the commitment
+	// state.
+	if deltas.Local >= 0 {
+		ourBalance += lnwire.MilliSatoshi(deltas.Local)
+	} else {
+		ourBalance -= lnwire.MilliSatoshi(-1 * deltas.Local)
+	}
+	if deltas.Remote >= 0 {
+		theirBalance += lnwire.MilliSatoshi(deltas.Remote)
+	} else {
+		theirBalance -= lnwire.MilliSatoshi(-1 * deltas.Remote)
+	}
+
+	if updateState {
+		state := lc.channelState
+		received := &state.TotalMSatReceived
+		sent := &state.TotalMSatSent
+
+		setHeights := func(u *paymentDescriptor) {
+			switch u.EntryType {
+			case Add:
+				u.addCommitHeights.SetForParty(
+					whoseCommitChain, nextHeight,
+				)
+			case Settle, Fail, MalformedFail:
+				u.removeCommitHeights.SetForParty(
+					whoseCommitChain, nextHeight,
+				)
+			case FeeUpdate:
+				u.addCommitHeights.SetForParty(
+					whoseCommitChain, nextHeight,
+				)
+				u.removeCommitHeights.SetForParty(
+					whoseCommitChain, nextHeight,
+				)
+			}
+		}
+
+		recordFlow := func(acc *lnwire.MilliSatoshi,
+			u *paymentDescriptor) {
+
+			if u.EntryType == Settle {
+				*acc += u.Amount
+			}
+		}
+
+		for _, u := range uncommitted.Local {
+			setHeights(u)
+			if whoseCommitChain == lntypes.Local {
+				recordFlow(received, u)
+			}
+		}
+		for _, u := range uncommitted.Remote {
+			setHeights(u)
+			if whoseCommitChain == lntypes.Local {
+				recordFlow(sent, u)
+			}
+		}
+	}
+
```

Comment:
Nit: [comment](https://github.com/lightningnetwork/lnd/pull/9097/files#diff-cdf9f47cef64ceac10ee55ed7bf0569ac54f5d5f1af599ef51ccaf5db813d03eR4658-R4661) needs a slight update, the underflow can happen now in computeView

Reply:
I don't know what comment is being referenced here. The link seems to go nowhere.

Reply:
```
	// We need to first check ourBalance and theirBalance to be negative
	// because MilliSathoshi is a unsigned type and can underflow in
	// `evaluateHTLCView`. This should never happen for views which do not
	// include new updates (remote or local).
```
in `computeView` func

the underflow can now only happen in `computeView` itself because we do the `ourBalance` and `theirBalance` there now.